README.md
# Loose JSON + TOML parsers for Go
Super loose JSON + TOML parsers and unmarshaller for Go.  
JSONC and JSON5 can also be read.

[![Test](https://github.com/shellyln/go-loose-json-parser/actions/workflows/test.yml/badge.svg)](https://github.com/shellyln/go-loose-json-parser/actions/workflows/test.yml)
[![release](https://img.shields.io/github/v/release/shellyln/go-loose-json-parser)](https://github.com/shellyln/go-loose-json-parser/releases)
[![Go version](https://img.shields.io/github/go-mod/go-version/shellyln/go-loose-json-parser)](https://github.com/shellyln/go-loose-json-parser)

<img src="https://raw.githubusercontent.com/shellyln/go-loose-json-parser/master/_assets/logo-go-jsonlp.svg" alt="logo" style="width:250px;" width="250">

---

## 📦 Example app

* [Loose JSON | TOML normalizer](https://shellyln.github.io/jsonlp/) (WebAssembly)
    * Normalize loose JSON | TOML to strict JSON
    * [Code](https://github.com/shellyln/go-loose-json-parser/blob/master/cmd/wasm/main.go)

## 🚀 Getting started

### JSON
```go
package main

import (
    "fmt"
    "github.com/shellyln/go-loose-json-parser/jsonlp"
)

func main() {
    // src: Loose JSON
    //
    // plafLb:
    // * Platform-dependent line break.
    //   * `Linebreak_Lf` | `Linebreak_CrLf` | `Linebreak_Cr`
    // * Line break codes in multi-line string are replaced by this specified line break.
    //   (Excluding line breaks by escape sequences)
    //
    // interop:
    // * If `Interop_JSON` is set,
    //   replace NaN, Infinity, complex number
    //   by `{nan:true}`, `{inf:+/-1}`, `{re:re,im:im}` .
    // * If `Interop_TOML` is set,
    //   replace complex number by `{re:re,im:im}` .
    // * If `Interop_JSON_AsNull` is set,
    //   replace NaN, Infinity, complex number by null.
    // * If `Interop_TOML_AsNull` is set,
    //   replace complex number by null.
    //
    // parsed:
    //   nil | []any | map[string]any |
    //   float64 | int64 | uint64 | complex128 |
    //   string | bool | time.Time
    parsed, err := jsonlp.ParseJSON(`{
        // comment
        config: {
            addr: '127.0.0.1',
        }
    }`, jsonlp.Linebreak_Lf, jsonlp.Interop_None)

    if err != nil {
        fmt.Printf("Parse: error = %v\n", err)
        return
    }

    fmt.Printf("Parsed = %v\n", parsed)
}
```

### TOML

```go
package main

import (
    "fmt"
    "github.com/shellyln/go-loose-json-parser/jsonlp"
)

func main() {
    // src: Loose TOML
    //
    // plafLb:
    // * Platform-dependent line break.
    //   * `Linebreak_Lf` | `Linebreak_CrLf` | `Linebreak_Cr`
    // * Line break codes in multi-line string are replaced by this specified line break.
    //   (Excluding line breaks by escape sequences)
    //
    // interop:
    // * If `Interop_JSON` is set,
    //   replace NaN, Infinity, complex number
    //   by `{nan:true}`, `{inf:+/-1}`, `{re:re,im:im}` .
    // * If `Interop_TOML` is set,
    //   replace complex number by `{re:re,im:im}` .
    // * If `Interop_JSON_AsNull` is set,
    //   replace NaN, Infinity, complex number by null.
    // * If `Interop_TOML_AsNull` is set,
    //   replace complex number by null.
    //
    // parsed:
    //   nil | []any | map[string]any |
    //   float64 | int64 | uint64 | complex128 |
    //   string | bool | time.Time
    parsed, err := jsonlp.ParseTOML(`
    # comment
    [config]
    addr = '127.0.0.1'
    `, jsonlp.Linebreak_Lf, jsonlp.Interop_None)

    if err != nil {
        fmt.Printf("Parse: error = %v\n", err)
        return
    }

    fmt.Printf("Parsed = %v\n", parsed)
}
```

### Unmarshal
Mapping untyped data to a typed variable.
```go
package main

import (
    "fmt"
    "github.com/shellyln/go-loose-json-parser/jsonlp"
    "github.com/shellyln/go-loose-json-parser/marshal"
)

type config struct {
    Addr string `json:"addr"`
}

type response struct {
    Config config `json:"config"`
}

func main() {
    parsed, err := jsonlp.ParseJSON(`{
        // comment
        config: {
            addr: '127.0.0.1',
        }
    }`, jsonlp.Linebreak_Lf, jsonlp.Interop_None)

    if err != nil {
        fmt.Printf("Parse: error = %v\n", err)
        return
    }

    var typed response

    // src:  Source data. Untyped in typical use.
    // dst:  Pointer to result data. Typed in typical use.
    // opts: Pointer to struct of the `Unmarshal` options. If nil, use default.
    //       Default options are {
    //           TagName: "json",               // Tag name of the struct fields
    //           NoCopyUnexportedFields: false, // If true, no shallow copying of unexported fields
    //           NoCustomMarshaller: false,     // If true, IMarshal and IUnmarshal are not used
    //       }
    if err := marshal.Unmarshal(parsed, &typed, nil); err != nil {
        fmt.Printf("Unmarshal: error = %v\n", err)
        return
    }

    fmt.Printf("Typed = %v\n", typed)
}

```

> **Note**  
> `Unmarshal` also works well for typed to untyped conversions and as deep cloning.

## 🥅 Goal
* ✅ Can read strict TOML.
* ✅ Can read loose JSON, JSONC, JSON5, and TOML for configuration files.
* ✅ Can be mapped to structs, untyped maps/slices, or primitives.
* ✅ Can read many list and map literals in JavaScript, Python, PHP, Ruby, etc.

## 🚧 ToDo

### APIs

* ✅ ~~Marshalling from any to typed~~

### TOML

* ✅ ~~Datetime format with date and time delimited by space (RFC 3339 section 5.6)~~
  * ~~e.g. `2006-01-02 15:04:05Z`~~
* ✅ ~~Datetime format without timezone~~
  * ~~e.g. `2006-01-02T15:04:05`~~
* ✅ ~~Platform-dependent newline in multiline string~~
* Error detection when values are overwritten


## 🪄 Examples

### JSON

```js
# Hash comment
{
    // Line comment
    /* Block comment */

    // Object keys can be enclosed in either double-quote, single-quote, back-quote.
    // It is also allowed not to enclose them.

    "foo": [
        123,          // -> float64(123)
        -123.45,      // -> float64(-123.45)
        -1.2345e+6,   // -> float64(-1234500)
        -123_456_789, // -> float64(-123456789)
        0x12345678,   // -> float64(305419896)
        0x1234_5678,  // -> float64(305419896)
        0o7654_3210,  // -> float64(16434824)
        0b0101_0101,  // -> float64(85)
    ],

    'bar': null,      // -> nil
    baz: undefined,   // -> nil
    "qux": -Infinity, // -> -Inf // Infinity, +Infinity are also available
    "quux": NaN,      // -> NaN

    // Non-ASCII identifiers can also be used.
    あいうえお: {
        key: "value1",
        bare_key: "value2",
        bare-key: "value3", // Keys containing hyphens are allowed.
        1234: "value4",     // Keys starting with a number are allowed.
        -3.14: "value5",    // "-3": { "14": "value5" }
    },

    "corge": [
        // Escape sequences are available
        'Hello, World!\r\n',

        // Unicode escape sequences (\uXXXX and \u{XXXXXX}) are available
        "\u0048\u0065\u006c\u006c\u006f\u002c\u0020\u0057\u006f\u0072\u006c\u0064\u0021",

        // Byte escape sequence is also available
        "\x48\x65\x6c\x6c\x6f\x2c\x20\x57\x6f\x72\x6c\x64\x21",

        // Multiline string literal
        `Tempor adipiscing amet velit ipsum diam ut ea takimata lorem gubergren sed laoreet.
        Congue possim facilisis sea justo dolore amet eos dolores est magna.`
    ],

    "grault": [
        // Date, Time, DateTime literals are available

        2020-12-31,              // -> time.Time 2020-12-31:00:00.000Z
        18:20:30.001,            // -> time.Time 1870-01-01T18:20:30.001Z
        2020-12-31T18:20:30.001Z // -> time.Time 2020-12-31:20:30.001Z
    ],

    // "key = value" syntax is allowed
    garply = true,

    // "key => value" syntax is allowed
    waldo => false,

    // Trailing commas are allowed.
    fred: 10,
}
```

See also: [Introducing JSON](https://www.json.org/json-en.html), [JSON5 Spec](https://spec.json5.org/)

### TOML

```toml
# Hash comment
// Line comment (non-standard)
/* Block comment (non-standard) */

"quoted-key" = "value0"
key          = "value1"
bare_key     = "value2"
bare-key     = "value3"  # Keys containing hyphens are allowed.
1234         = "value4"  # Keys starting with a number are allowed.
-3.14        = "value5"  # "-3": { "14": "value5" }
あいうえお    = "value6"  # // Non-ASCII identifiers are allowed. (non-standard)

[table-A]
item-a = 1
item-b = 2
item-c = "standard\n string"
item-d = '\literal string'
item-e = """\
         Multiline \
         standard string"""
item-f = '''
Multiline
literal string''' "\t" '''!!'''

[table-A.sub-table-P]
item-m = 11
item-n = 12
sub-table-X.sub-table-Y = 111

[[array-of-table-U]]
foo = 1
[array-of-table-U.sub-table-V]
bar = 11

[[array-of-table-U]]
foo = 2
[array-of-table-U.sub-table-V]
bar = 22
```

See also: [TOML Spec](https://toml.io/en/v1.0.0)


## 📚 Syntax

### Array

```js
[]
```

```js
[1]
```

```js
[1,]
```

```js
[1, 2]
```

```js
[1, 2,]
```

### Object

```js
{}
```

```js
{ "foo": 1 }
```

```js
{ "foo": 1, }
```

```js
{ 'foo': 1, }
```

```js
{ `foo`: 1, }
```

```js
{ foo: 1, }
```

```js
{ "foo": 1, "bar": 2, }
```

```js
{ "foo" => 1 }
```

```js
{ "foo" = 1 }
```

```js
{ foo.bar.baz = 1 }
// -> { "foo": { "bar": { "baz": 1 } } }
```

```js
{ "foo".bar."baz" = 1 }
// -> { "foo": { "bar": { "baz": 1 } } }
```

### Number

```js
123
```
```js
-123
```
```js
-123s64
```
```js
123u64
```
```js
-123.45
```
```js
-1.2345e+6
```
```js
-123_456_789
```
```js
-123_456.789_012
```
```js
0x12345678
```
```js
0x1234_5678
```
```js
0o7654_3210
```
```js
0b0101_0101
```
```js
0x1p-2
```
```js
0x1.Fp+0
```
```js
0X_1_FFF_P-16
```
```js
NaN
```
```js
nan
```
```js
Infinity
```
```js
+Infinity
```
```js
-Infinity
```
```js
inf
```
```js
+inf
```
```js
-inf
```

### Complex

```js
1.23 - 34.5i
```
```js
1.23e+1 - 34.5e-1i
```
```js
0x1.8p+1 - 0x1.8p-1i
```
```js
NaN-NaNi
```
```js
NaN - NaN_i
```
```js
Infinity-Infinityi
```
```js
Infinity - Infinity_i
```

### String

```js
"foobar"
```
```js
'foobar'
```
```js
`foo
bar`
```
```yaml
"""foo
bar"""
```
```yaml
'''foo
bar'''
```
```js
"Hello\n\r\v\t\b\fWorld!"
```
```js
"\u0048\u0065\u006c\u006c\u006f\u002c\u0020\u0057\u006f\u0072\u006c\u0064\u0021"
```
```js
"\u{000048}\u{000065}\u{00006c}\u{00006c}\u{00006f}\u{00002c}\u{000020}\u{000057}\u{00006f}\u{000072}\u{00006c}\u{000064}\u{000021}"
```
```js
"\x48\x65\x6c\x6c\x6f\x2c\x20\x57\x6f\x72\x6c\x64\x21"
```

```js
"foo" "bar"
// -> "foobar"
```
```js
'foo' 'bar'
// -> 'foobar'
```

### Boolean

```js
true
```
```js
false
```

### Null, Undefined

```js
null
// -> nil
```
```js
undefined
// -> nil
```
```rust
None
// -> nil
```

### Comment

```sh
# Hash line comment
```
```js
// Line comment
```
```js
/*
   Block comment
*/
```

## ⚖️ License

MIT  
Copyright (c) 2023 Shellyl_N and Authors.


---
Powered by [takenoco](https://github.com/shellyln/takenoco) Parser Combinator Library


Changelog.md
# Changelog

# v0.0.19
* Edit package comments.

# v0.0.18
* Added support for platform-dependent line breaks.
* [Breaking change] Change function name `Parse` -> `ParseJSON`.
* [Breaking change] Change `ParseJSON`, `ParseTOML` signatures.

# v0.0.17
* Added Datetime formats.
  * Datetime format with date and time delimited by space (RFC 3339 section 5.6). e.g. `2006-01-02 15:04:05Z`
  * Datetime format without timezone. e.g. `2006-01-02T15:04:05`

# v0.0.16
* Empty string to number marshalling sets a zero value.
* Changed some keywords to be case-insensitive.
  * `null`, `inf`, `nan`
* Added keyword `None`.
* Edit README.

# v0.0.15
* Using constants for ClassName.
* Edit README.

# v0.0.14
* [FIX] Fix unmarshaller; Conversion from map to struct causes an error if there is no key matching the field name in the map.

# v0.0.13
* Added unmarshaller.
* [Breaking change] The required minimum version of Go has changed from 1.17 to 1.18.
  * Change `go.mod` and `test.yml`.
* Edit README.
* Edit wasm example html.

# v0.0.12
* [Breaking change] Change interop parameter type.
* Edit README.

# v0.0.11
* Added complex number parser.
* Remove unnecessary `erase()` call.
* Added dummy `version.go` file: for calling `go get -u`.
* Edit README.

# v0.0.10
* Improve word boundary.
* Edit wasm example html.
* Edit README.

# v0.0.9
* Improve JSON string parser: support TOML style multiline strings.
* Edit wasm example.
* Edit README.

# v0.0.8
* Added hexadecimal float literals.
* Added tests.
* Edit wasm example html.

# v0.0.7
* Fix string parsers: escape sequence.
* Added strict IEEE-754 +0/-0 parsing test.
* Edit README.

# v0.0.6
* Added TOML parser.
* Edit README.
* Rename files.

# v0.0.5
* Added dotted key syntax.

# v0.0.4
* Added '=' to the key/value separator in maps.

# v0.0.3
* Add wasm example.

# v0.0.2
* (Breaking change) Change `Parse` parameters.
  * Add `interop` parameter.
* Improve error reporting.
* ~~Add wasm example.~~
* Edit README.

# v0.0.1
* First release.


jsonlp/parser.go
package jsonlp

import (
	"errors"
	"strconv"
	"time"

	. "github.com/shellyln/takenoco/base"
	"github.com/shellyln/takenoco/extra"
	. "github.com/shellyln/takenoco/string"
)

type parseOptions struct {
	interop bool
}

var (
	documentParser ParserFn
)

func init() {
	documentParser = document()
}

// Remove the resulting AST.
func erase(fn ParserFn) ParserFn {
	return Trans(fn, Erase)
}

// Whitespaces
func sp0() ParserFn {
	return erase(ZeroOrMoreTimes(First(Whitespace(), comment())))
}

func lineComment() ParserFn {
	return erase(FlatGroup(
		Seq("//"),
		FlatGroup(
			ZeroOrMoreTimes(CharClassN("\r\n", "\n", "\r")),
			First(
				CharClass("\r\n", "\n", "\r"),
				LookAhead(End()),
			),
		),
	))
}

func hashLineComment() ParserFn {
	return erase(FlatGroup(
		Seq("#"),
		FlatGroup(
			ZeroOrMoreTimes(CharClassN("\r\n", "\n", "\r")),
			First(
				CharClass("\r\n", "\n", "\r"),
				LookAhead(End()),
			),
		),
	))
}

func blockComment() ParserFn {
	return erase(FlatGroup(
		Seq("/*"),
		ZeroOrMoreTimes(CharClassN("*/")),
		First(
			Seq("*/"),
			Error("An unexpected termination has appeared in the block comment."),
		),
	))
}

func comment() ParserFn {
	return First(lineComment(), hashLineComment(), blockComment())
}

func trueValue() ParserFn {
	return FlatGroup(
		erase(Seq("true")),
		WordBoundary(),
		Zero(trueAst),
	)
}

func falseValue() ParserFn {
	return FlatGroup(
		erase(Seq("false")),
		WordBoundary(),
		Zero(falseAst),
	)
}

func boolValue() ParserFn {
	return First(
		trueValue(),
		falseValue(),
	)
}

func nullValue() ParserFn {
	return FlatGroup(
		erase(CharClass("null", "undefined")),
		WordBoundary(),
		Zero(nilAst),
	)
}

func positiveInfinityValue() ParserFn {
	return FlatGroup(
		erase(FlatGroup(
			ZeroOrOnce(Seq("+")),
			Seq("Infinity"),
		)),
		WordBoundary(),
		func(ctx ParserContext) (ParserContext, error) {
			if ctx.Tag.(parseOptions).interop {
				return Zero(nilAst)(ctx)
			} else {
				return Zero(positiveInfinityAst)(ctx)
			}
		},
	)
}

func negativeInfinityValue() ParserFn {
	return FlatGroup(
		erase(Seq("-Infinity")),
		WordBoundary(),
		func(ctx ParserContext) (ParserContext, error) {
			if ctx.Tag.(parseOptions).interop {
				return Zero(nilAst)(ctx)
			} else {
				return Zero(negativeInfinityAst)(ctx)
			}
		},
	)
}

func nanValue() ParserFn {
	return FlatGroup(
		erase(Seq("NaN")),
		WordBoundary(),
		func(ctx ParserContext) (ParserContext, error) {
			if ctx.Tag.(parseOptions).interop {
				return Zero(nilAst)(ctx)
			} else {
				return Zero(nanAst)(ctx)
			}
		},
	)
}

func numberValue() ParserFn {
	return First(
		FlatGroup(
			First(
				Trans(
					FlatGroup(erase(SeqI("0b")), extra.BinaryNumberStr()),
					ParseIntRadix(2),
					castIntToFloat,
				),
				Trans(
					FlatGroup(erase(SeqI("0o")), extra.OctalNumberStr()),
					ParseIntRadix(8),
					castIntToFloat,
				),
				Trans(
					FlatGroup(erase(SeqI("0x")), extra.HexNumberStr()),
					ParseIntRadix(16),
					castIntToFloat,
				),
				Trans(
					extra.FloatNumberStr(),
					ParseFloat,
					ChangeClassName("Float"),
				),
				Trans(
					extra.IntegerNumberStr(),
					ParseInt,
					castIntToFloat,
				),
			),
			WordBoundary(),
		),
		positiveInfinityValue(),
		negativeInfinityValue(),
		nanValue(),
	)
}

func stringLiteralInner(cc string, multiline bool) ParserFn {
	return FlatGroup(
		erase(Seq(cc)),
		ZeroOrMoreTimes(
			First(
				FlatGroup(
					erase(Seq("\\")),
					First(
						CharClass("\\", "'", "\"", "`"),
						replaceStr(CharClass("n", "N"), "\n"),
						replaceStr(CharClass("r", "R"), "\r"),
						replaceStr(CharClass("v", "V"), "\v"),
						replaceStr(CharClass("t", "T"), "\t"),
						replaceStr(CharClass("b", "B"), "\b"),
						replaceStr(CharClass("f", "F"), "\f"),
						Trans(
							FlatGroup(
								erase(CharClass("u")),
								Repeat(Times{Min: 4, Max: 4}, HexNumber()),
							),
							ParseIntRadix(16),
							StringFromInt,
						),
						Trans(
							FlatGroup(
								erase(CharClass("u{")),
								Repeat(Times{Min: 1, Max: 6}, HexNumber()),
								erase(CharClass("}")),
							),
							ParseIntRadix(16),
							StringFromInt,
						),
						Trans(
							FlatGroup(
								erase(CharClass("x")),
								Repeat(Times{Min: 2, Max: 2}, HexNumber()),
							),
							ParseIntRadix(16),
							StringFromInt,
						),
						Trans(
							FlatGroup(
								Repeat(Times{Min: 3, Max: 3}, OctNumber()),
							),
							ParseIntRadix(8),
							StringFromInt,
						),
					),
				),
				If(multiline,
					OneOrMoreTimes(CharClassN(cc, "\\")),
					OneOrMoreTimes(
						First(
							FlatGroup(
								CharClass("\r", "\n"),
								Error("An unexpected newline has appeared in the string literal."),
							),
							CharClassN(cc, "\\"),
						),
					),
				),
			),
		),
		First(
			FlatGroup(End(), Error("An unexpected termination has appeared in the string literal.")),
			erase(Seq(cc)),
		),
	)
}

func stringValue() ParserFn {
	return Trans(
		First(
			stringLiteralInner("\"", false),
			stringLiteralInner("'", false),
			stringLiteralInner("`", true),
		),
		Concat,
		ChangeClassName("String"),
	)
}

func dateValue() ParserFn {
	return Trans(
		FlatGroup(
			extra.DateStr(),
			WordBoundary(),
		),
		func(ctx ParserContext, asts AstSlice) (AstSlice, error) {
			value := asts[len(asts)-1].Value.(string)
			t, err := time.Parse("2006-01-02", value)
			if err != nil {
				return nil, err
			}
			return AstSlice{{
				ClassName: "Date",
				Type:      AstType_Any,
				Value:     t.UTC(),
			}}, nil
		},
	)
}

func dateTimeValue() ParserFn {
	return Trans(
		FlatGroup(
			extra.DateTimeStr(),
			WordBoundary(),
		),
		func(ctx ParserContext, asts AstSlice) (AstSlice, error) {
			// TODO: BUG: Cannot parse years with negative values or years greater than or equal to 10000.
			value := asts[len(asts)-1].Value.(string)
			t, err := time.Parse("2006-01-02T15:04:05.000000000-07:00", value)
			if err != nil {
				return nil, err
			}
			return AstSlice{{
				ClassName: "DateTime",
				Type:      AstType_Any,
				Value:     t.UTC(),
			}}, nil
		},
	)
}

func timeValue() ParserFn {
	return Trans(
		FlatGroup(
			extra.TimeStr(),
			WordBoundary(),
		),
		func(ctx ParserContext, asts AstSlice) (AstSlice, error) {
			value := "1970-01-01T" + asts[len(asts)-1].Value.(string) + "+00:00"
			t, err := time.Parse("2006-01-02T15:04:05.000000000-07:00", value)
			if err != nil {
				return nil, err
			}
			return AstSlice{{
				ClassName: "Time",
				Type:      AstType_Any,
				Value:     t.UTC(),
			}}, nil
		},
	)
}

func primitiveValue() ParserFn {
	return First(
		stringValue(),
		boolValue(),
		nullValue(),
		timeValue(),
		dateTimeValue(),
		dateValue(),
		numberValue(),
	)
}

func listValue() ParserFn {
	return Trans(
		FlatGroup(
			erase((Seq("["))),
			erase(sp0()),
			ZeroOrOnce(
				FlatGroup(
					First(
						primitiveValue(),
						Indirect(listValue),
						Indirect(objectValue),
					),
					erase(sp0()),
				),
				ZeroOrMoreTimes(
					erase((Seq(","))),
					erase(sp0()),
					First(
						primitiveValue(),
						Indirect(listValue),
						Indirect(objectValue),
						LookAhead(Seq("]")),
						FlatGroup(
							erase(sp0()),
							Error("Expect array closing parenthesis ')' or value"),
						),
					),
					erase(sp0()),
				),
			),
			ZeroOrOnce(
				erase((Seq(","))),
				erase(sp0()),
			),
			First(
				erase((Seq("]"))),
				FlatGroup(
					erase(sp0()),
					Error("Expect array closing parenthesis ')'"),
				),
			),
			erase(sp0()),
		),
		func(ctx ParserContext, asts AstSlice) (AstSlice, error) {
			length := len(asts)
			v := make([]interface{}, length)
			for i := 0; i < length; i++ {
				v[i] = asts[i].Value
			}
			return AstSlice{{
				ClassName: "Array",
				Type:      AstType_Any,
				Value:     v,
			}}, nil
		},
	)
}

func identifier() ParserFn {
	return Trans(
		extra.UnicodeIdentifierStr(),
		ChangeClassName("Idenitifier"),
	)
}

func objectKey() ParserFn {
	return Trans(
		First(
			stringValue(),
			identifier(),
		),
	)
}

func objectKeyValuePair() ParserFn {
	return FlatGroup(
		objectKey(),
		sp0(),
		erase(First(CharClass(":"), CharClass("=>"), CharClass("="))),
		sp0(),
		First(
			primitiveValue(),
			Indirect(listValue),
			Indirect(objectValue),
			Error("Expect object property value"),
		),
		sp0(),
	)
}

func objectValue() ParserFn {
	return Trans(
		FlatGroup(
			erase((Seq("{"))),
			erase(sp0()),
			ZeroOrOnce(
				objectKeyValuePair(),
				ZeroOrMoreTimes(
					erase((Seq(","))),
					erase(sp0()),
					First(
						objectKeyValuePair(),
						LookAhead(Seq("}")),
						FlatGroup(
							erase(sp0()),
							Error("Expect object closing bracket '}' or key-value pair"),
						),
					),
				),
				ZeroOrOnce(
					erase((Seq(","))),
					erase(sp0()),
				),
			),
			First(
				erase((Seq("}"))),
				FlatGroup(
					erase(sp0()),
					Error("Expect object closing bracket '}'"),
				),
			),
			erase(sp0()),
		),
		func(ctx ParserContext, asts AstSlice) (AstSlice, error) {
			length := len(asts)
			v := make(map[string]interface{})
			for i := 0; i < length; i += 2 {
				v[asts[i].Value.(string)] = asts[i+1].Value
			}
			return AstSlice{{
				ClassName: "Object",
				Type:      AstType_Any,
				Value:     v,
			}}, nil
		},
	)
}

func document() ParserFn {
	return FlatGroup(
		Start(),
		erase(sp0()),
		First(
			primitiveValue(),
			listValue(),
			objectValue(),
			erase(sp0()),
		),
		First(
			End(),
			Error("Expect terminatiion"),
		),
	)
}

// src:     Loose JSON
// interop: If true, replace NaN, Infinity by null
// parsed:  nil | []any | map[string]any | float64 | string | bool | time.Time
func Parse(s string, interop bool) (interface{}, error) {
	ctx := *NewStringParserContext(s)
	ctx.Tag = parseOptions{interop: interop}

	out, err := documentParser(ctx)
	if err != nil {
		pos := GetLineAndColPosition(s, out.SourcePosition, 4)
		return nil, errors.New(
			err.Error() +
				"\n --> Line " + strconv.Itoa(pos.Line) +
				", Col " + strconv.Itoa(pos.Col) + "\n" +
				pos.ErrSource)
	}

	if out.MatchStatus == MatchStatus_Matched {
		return out.AstStack[0].Value, nil
	} else {
		pos := GetLineAndColPosition(s, out.SourcePosition, 4)
		return nil, errors.New(
			"Parse failed" +
				"\n --> Line " + strconv.Itoa(pos.Line) +
				", Col " + strconv.Itoa(pos.Col) + "\n" +
				pos.ErrSource)
	}
}
